### DISCUSSION TITLE
Unstable client connection to a Sign&Encrypt server

### DISCUSSION THREAD

**Discussion by danmoser** [FreeOpcUa/opcua-asyncio#1237] [Q&A] [answered]
Hi! I have a asyncua server using `ua.SecurityPolicyType.Basic256Sha256_SignAndEncrypt`. The certificate was generated as in the [example](https://github.com/FreeOpcUa/opcua-asyncio/blob/master/examples/generate_certificate.sh).

The problem I'm facing is the following: if an open62541 client losses connection to that server (eg., due to an unstable network connection), it triggers "BadConnectionClosed" and "BadSecurityChecksFailes" errors, and the client dies. A "Basic" security policy doesn't trigger the errors.

I'm not sure if this is related, but every time I connect to the server with UaExpert, I get two warnings (as below): "BadCertificateUseNotAllowed" and "BadCertificateUriInvalid". After ignoring the two warnings, a connection is established. I noticed that during the network glitches, UaExpert also triggers "BadConnectionClosed" and "BadSessionIdInvalid" errors in its log. The two warnings are shown again, and after ignoring them the connection is usually re-established. Sometimes it's not, and I need to manually disconnect and reconnect to resume communicating with the server.

Do you have any recommendations on how to make the client connection more robust against the glitches?
Thanks in advance

 
![BadCertUseNotAllowed](https://user-images.githubusercontent.com/2973004/222858622-ace4fdda-810d-4364-afe9-04bd7056f8d3.png)
![BadCertUriInvalid](https://user-images.githubusercontent.com/2973004/222858626-1cbc0e5a-05b7-4202-ab27-88df2ea0e8da.png)

**Comment by schroeder-** [comment, marked as answer]
Did you change your application uri? Because the uri must match urn:opcua:python:server or you must change the alternate name uri parameter in the script.

**Reply by danmoser** [reply]
@schroeder- , after replacing the URI to `urn:freeopcua:python:server` the "BadCertificateUriInvalid" is gone. However, the "BadCertificateUseNotAllowed" persists:
![Screenshot_2023-03-06_08-27-54](https://user-images.githubusercontent.com/2973004/223156044-1e80603c-19c4-491c-a562-6f903b2f49ba.png)

**Reply by danmoser** [reply]
Just to reinforce that my main problem is that after the connection glitches, the client is not able to re-establish the connection with the server automatically (that's when this message appears). The connection is only re-established by manually disconnecting and reconnecting.

**Reply by schroeder-** [reply]
Is the certificate in the uaexpert trust list under Settings-> Manage Certificates?

**Reply by danmoser** [reply]
Yes, it is. And the problem also happens using an open62541 client -- but I haven't experienced that (yet) with an asyncua client.

**Reply by AndreasHeine** [reply]
have you checked whats in the keyusage field of the cert?

**Reply by schroeder-** [reply]
Maybe there is missing some extension on the certificate:
<html>
<body>


keyUsage | keyUsage | Specifies how the Certificate key may be used.Shall include digitalSignature, nonRepudiation, keyEncipherment and dataEncipherment.Other key uses are allowed.
-- | -- | --
extendedKeyUsage | keyUsage | Specifies additional key uses for the Certificate.Shall specify ‘serverAuth and/or clientAuth.Other key uses are allowed.


</body>
</html>

**Reply by AndreasHeine** [reply]
i wrote a how to in 2020:

Generate your own x509v3 Certificate
Step 1: Change ssl.conf (subjectAltname, country, organizationName, ...)
ssl.conf:
```
[ req ]
default_bits = 2048
default_md = sha256
distinguished_name = subject
req_extensions = req_ext
x509_extensions = req_ext
string_mask = utf8only
prompt = no
[ req_ext ]
basicConstraints = CA:FALSE
nsCertType = client, server
keyUsage = nonRepudiation, digitalSignature, keyEncipherment, dataEncipherment, keyCertSign
extendedKeyUsage= serverAuth, clientAuth
nsComment = "OpenSSL Generated Certificat"
subjectKeyIdentifier=hash
authorityKeyIdentifier=keyid,issuer
subjectAltName = URI:urn:opcua:python:server,IP: 127.0.0.1
[ subject ]
countryName = DE
stateOrProvinceName = HE
localityName = HE
organizationName = AndreasHeine
commonName = PythonOpcUaServer
```
Step 2: `openssl genrsa -out key.pem 2048`
Step 3: `openssl req -x509 -days 365 -new -out certificate.pem -key key.pem -config ssl.conf`
this way is proved with Siemens OPC UA Client/Server!

**Reply by danmoser** [reply]
I believe using the `ssl.conf` file solved it! Thank you!!!!!
